Compile errors

I'm getting the following output when trying to build the package with `gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)`. I also tried `gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)` on a different machine and the output is similar. Should I be using a different gcc version?

```
gcc -pthread -B /tsi/doctorants/ocifka/envs/style_rank/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC [1526/4209]
orants/ocifka/envs/style_rank/include -I/tsi/doctorants/ocifka/envs/style_rank/include -I/tsi/doctorants/ocifka/envs/style_rank/include/python3.6m -c src/style_rank/bind
ings.cpp -o build/temp.linux-x86_64-3.6/src/style_rank/bindings.o -DVERSION_INFO="1.0.19" -std=c++14 -fvisibility=hidden                                                 
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++                                                                          
In file included from src/style_rank/bindings.cpp:1:0:                                                                                                                   
src/style_rank/utils.hpp:113:39: error: ‘unique_ptr’ is not a member of ‘std’                                                                                            
     std::map<std::string, std::vector<std::unique_ptr<DISCRETE_DIST>>> dists;                                                                                           
                                       ^~~                                                                                                                               
src/style_rank/utils.hpp:113:39: error: ‘unique_ptr’ is not a member of ‘std’                                                                                            
src/style_rank/utils.hpp:113:55: error: template argument 1 is invalid                                                                                                   
     std::map<std::string, std::vector<std::unique_ptr<DISCRETE_DIST>>> dists;                                                                                           
                                                       ^~~~~~~~~~~~~                                                                                                     
src/style_rank/utils.hpp:113:55: error: template argument 2 is invalid                                                                                                   
src/style_rank/utils.hpp:113:68: error: template argument 2 is invalid
     std::map<std::string, std::vector<std::unique_ptr<DISCRETE_DIST>>> dists;
                                                                    ^~
src/style_rank/utils.hpp:113:68: error: template argument 4 is invalid
src/style_rank/utils.hpp:113:70: error: expected unqualified-id before ‘>’ token
     std::map<std::string, std::vector<std::unique_ptr<DISCRETE_DIST>>> dists;
                                                                      ^
src/style_rank/utils.hpp:116:37: error: ‘std::unique_ptr’ has not been declared
     void add(std::string name, std::unique_ptr<DISCRETE_DIST> x) {
                                     ^~~~~~~~~~
src/style_rank/utils.hpp:116:47: error: expected ‘,’ or ‘...’ before ‘<’ token
     void add(std::string name, std::unique_ptr<DISCRETE_DIST> x) {
                                               ^
src/style_rank/utils.hpp: In member function ‘void Collector::add(std::__cxx11::string, int)’:
src/style_rank/utils.hpp:117:32: error: ‘x’ was not declared in this scope
         for (const auto &kv : *x) {
                                ^
src/style_rank/utils.hpp:120:9: error: ‘dists’ was not declared in this scope
         dists[name].push_back(std::move(x));
         ^~~~~
src/style_rank/utils.hpp:120:41: error: ‘x’ was not declared in this scope
         dists[name].push_back(std::move(x));
                                         ^
src/style_rank/utils.hpp: In member function ‘std::tuple<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector[1490/4209]
ned int, std::allocator<long unsigned int> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const
 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > > >, std::map<std:
:__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<long unsigned int, std::allocator<long unsigned int> >, std::less<std::__cxx11::
basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocat
or<char> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > > > > Collector::getData(size_t)’:
src/style_rank/utils.hpp:131:31: error: ‘dists’ was not declared in this scope
         for (auto const &kv : dists) {
                               ^~~~~
src/style_rank/utils.hpp:147:38: error: unable to deduce ‘auto&&’ from ‘domain’
                 for (const auto &d : domain) {
                                      ^~~~~~
In file included from src/style_rank/bindings.cpp:2:0:
src/style_rank/parse.hpp: At global scope:
src/style_rank/parse.hpp:81:16: error: ‘unique_ptr’ was not declared in this scope
   PCINT(vector<unique_ptr<NOTE>>::iterator b, vector<unique_ptr<NOTE>>::iterator e) {
                ^~~~~~~~~~
src/style_rank/parse.hpp:81:27: error: template argument 1 is invalid
   PCINT(vector<unique_ptr<NOTE>>::iterator b, vector<unique_ptr<NOTE>>::iterator e) {
                           ^~~~
src/style_rank/parse.hpp:81:27: error: template argument 2 is invalid
src/style_rank/parse.hpp:81:31: error: expected ‘,’ or ‘...’ before ‘>’ token
   PCINT(vector<unique_ptr<NOTE>>::iterator b, vector<unique_ptr<NOTE>>::iterator e) {
                               ^~
src/style_rank/parse.hpp: In constructor ‘PCINT::PCINT(int)’:
src/style_rank/parse.hpp:83:20: error: ‘b’ was not declared in this scope
     for (auto it = b; it != e; it++) {
                    ^
src/style_rank/parse.hpp:83:29: error: ‘e’ was not declared in this scope
     for (auto it = b; it != e; it++) {
                             ^
src/style_rank/parse.hpp: At global scope:
src/style_rank/parse.hpp:106:10: error: ‘unique_ptr’ was not declared in this scope
   vector<unique_ptr<NOTE>> notes;
          ^~~~~~~~~~
src/style_rank/parse.hpp:106:21: error: template argument 1 is invalid
   vector<unique_ptr<NOTE>> notes;
                     ^~~~                                                                                                                                     [1453/4209]
src/style_rank/parse.hpp:106:21: error: template argument 2 is invalid
src/style_rank/parse.hpp:106:25: error: expected unqualified-id before ‘>’ token
   vector<unique_ptr<NOTE>> notes;
                         ^~
src/style_rank/parse.hpp: In member function ‘void Piece::addNote(int, int, int, int)’:
src/style_rank/parse.hpp:170:5: error: ‘notes’ was not declared in this scope
     notes.push_back(
     ^~~~~
src/style_rank/parse.hpp:171:7: error: ‘unique_ptr’ was not declared in this scope
       unique_ptr<NOTE>(new NOTE(pitch, onset, duration, velocity)));
       ^~~~~~~~~~
src/style_rank/parse.hpp:171:22: error: expected primary-expression before ‘>’ token
       unique_ptr<NOTE>(new NOTE(pitch, onset, duration, velocity)));
                      ^
src/style_rank/parse.hpp: In member function ‘void Piece::findChords(bool)’:
src/style_rank/parse.hpp:179:9: error: ‘notes’ was not declared in this scope
     if (notes.size() <= 0) return;
         ^~~~~
src/style_rank/parse.hpp:181:29: error: ‘notes’ was not declared in this scope
     for (const auto &note : notes) {
                             ^~~~~
In file included from src/style_rank/bindings.cpp:3:0:
src/style_rank/features.hpp: At global scope:
src/style_rank/features.hpp:30:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> IntervalDist(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:42:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> IntervalClassDist(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:54:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordSize(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:65:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordPCSizeRatio(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:84:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetRatio(Piece *p) /*ORIGINAL*/ {                                                                                           [1416/4209]
 ^~~~~~~~~~
src/style_rank/features.hpp:100:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordDistinctDurationRatio(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:115:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordDuration(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:126:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordShape(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:143:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetShape(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:163:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordPCD(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:174:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordPCDWBass(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:185:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetPCD(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:196:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetTiePCD(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:204:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetTiePCDTogether(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:222:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTonnetz(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:233:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnset(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:250:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordRange(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~                                                                                                                                                   [1379/4209]
src/style_rank/features.hpp:263:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordDissonance(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:288:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranDissonance(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:313:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordLowestInterval(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:326:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordSizeNgram(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:341:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranVoiceMotion(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:367:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranRepeat(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:392:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranScaleDistance(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:412:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranScaleUnion(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:432:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranDistance(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:445:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranOuter(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
In file included from src/style_rank/bindings.cpp:3:0:
src/style_rank/features.hpp:461:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranBassInterval(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:478:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordTranMelodyInterval(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~                                                                                                                                                   [1342/4209]
src/style_rank/features.hpp:496:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordMelodyNgram(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp: In function ‘uint64_t roll_to_min(uint64_t, int)’:
src/style_rank/features.hpp:513:29: warning: suggest parentheses around ‘+’ in operand of ‘&’ [-Wparentheses]
     uint64_t tmp = (x >> i) + (x << (n-i)) & ((1<<n)-1);
                    ~~~~~~~~~^~~~~~~~~~~~~~
src/style_rank/features.hpp: At global scope:
src/style_rank/features.hpp:521:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> PCDTran(Piece *p) /*ORIGINAL*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:539:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordSizeDurationWeighted(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:551:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> OffsetDistrubution(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:563:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> MelodicInterval(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:575:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> DurationDifference(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:587:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> OnsetDifference(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:599:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> Onset(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:611:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> Duration(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:623:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> MelodicNGramPCD(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:638:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordDurationMirex(Piece *p) /*MIREX*/ {                                                                                           [1305/4209]
 ^~~~~~~~~~
src/style_rank/features.hpp:658:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOnsetDifference(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:670:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> Pitch(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:724:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordOuterInterval(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
src/style_rank/features.hpp:744:1: error: ‘unique_ptr’ does not name a type
 unique_ptr<DISCRETE_DIST> ChordDistance(Piece *p) /*MIREX*/ {
 ^~~~~~~~~~
In file included from src/style_rank/bindings.cpp:4:0:
src/style_rank/feature_map.hpp:12:30: error: ‘unique_ptr’ was not declared in this scope
 static unordered_map<string, unique_ptr<DISCRETE_DIST>(*)(Piece*)> m {
                              ^~~~~~~~~~
src/style_rank/feature_map.hpp:12:54: error: template argument 2 is invalid
 static unordered_map<string, unique_ptr<DISCRETE_DIST>(*)(Piece*)> m {
                                                      ^
src/style_rank/feature_map.hpp:12:54: error: template argument 5 is invalid
src/style_rank/feature_map.hpp:12:57: error: expected unqualified-id before ‘)’ token
 static unordered_map<string, unique_ptr<DISCRETE_DIST>(*)(Piece*)> m {
                                                         ^
src/style_rank/feature_map.hpp:12:66: error: expected initializer before ‘>’ token
 static unordered_map<string, unique_ptr<DISCRETE_DIST>(*)(Piece*)> m {
                                                                  ^
In file included from src/style_rank/bindings.cpp:10:0:
/tsi/doctorants/ocifka/envs/style_rank/include/python3.6m/pybind11/pybind11.h:33:30: error: expected declaration before end of line
 #  pragma GCC diagnostic push
                              ^
error: command 'gcc' failed with exit status 1
```

--- a/src/style_rank/utils.hpp
+++ b/src/style_rank/utils.hpp
@@ -1,6 +1,7 @@
 #ifndef STYLE_RANK_UTILS_H
 #define STYLE_RANK_UTILS_H
 
+#include <memory>
 #include <vector>
 #include <numeric>
 #include <sstream>

